File: backend/functions/routes/chatRoutes.js
/**
 * backend/functions/routes/chatRoutes.js
 * Express routes for chat and message handling
 */

const express = require("express");
const router = express.Router();
const {
  ChatHistoryService,
  PersonalizationService,
  FirestoreService,
} = require("../services");
const logger = require("../utils/logger");

// Initialize services
const chatHistoryService = new ChatHistoryService();
const personalizationService = new PersonalizationService();
const firestoreService = new FirestoreService();

/**
 * POST /api/chat/:userId/message
 * Save a new message and process through RAG/Web/Hybrid modes
 */
router.post("/api/chat/:userId/message", async (req, res) => {
  try {
    const { userId } = req.params;
    const { message, conversationId, responseMode } = req.body;

    // Verify user is sending their own message
    if (req.user.uid !== userId) {
      return res
        .status(403)
        .json({ error: "Cannot send message as another user" });
    }

    // Validate required fields
    if (!message || !message.content) {
      return res.status(400).json({ error: "Message content is required" });
    }

    // Create message object
    const userMessage = {
      conversationId: conversationId || `conv_${Date.now()}`,
      sender: "user",
      content: message.content,
      timestamp: new Date().toISOString(),
      metadata: message.metadata || {},
    };

    // Save user message to Firestore
    await chatHistoryService.saveMessage(userId, userMessage);

    // Get user's settings for response mode preference
    const userSettings = await firestoreService.getUserSettings(userId);
    const mode = responseMode || userSettings?.responseMode || "hybrid";

    // Build personalization context for better response
    const personalizationContext =
      await personalizationService.buildPersonalizationContext(userId);

    logger.debug(`Processing message for user ${userId} in ${mode} mode`);

    // TODO: Integrate with your RAG system
    // This is where your answer_question() function should be called
    // Pass: message.content, mode, personalizationContext
    // Expected response: { answer, mode, sources }

    const response = {
      conversationId: userMessage.conversationId,
      mode,
      status: "pending", // Will be 'success' after RAG integration
      message: "Message received. RAG processing pending.",
    };

    res.status(200).json(response);
  } catch (error) {
    logger.error(`Error processing message: ${error.message}`);
    res.status(500).json({ error: "Error processing message" });
  }
});

/**
 * GET /api/chat/:userId/history
 * Get user's chat history
 */
router.get("/api/chat/:userId/history", async (req, res) => {
  try {
    const { userId } = req.params;
    const { limit = 50, offset = 0 } = req.query;

    // Verify user is accessing their own history
    if (req.user.uid !== userId) {
      return res
        .status(403)
        .json({ error: "Cannot access other user's chat history" });
    }

    const messages = await chatHistoryService.getUserChatHistory(
      userId,
      parseInt(limit),
      parseInt(offset),
    );

    res.status(200).json({ messages, count: messages.length });
  } catch (error) {
    logger.error(`Error retrieving chat history: ${error.message}`);
    res.status(500).json({ error: "Error retrieving chat history" });
  }
});

/**
 * GET /api/chat/:userId/conversations
 * Get user's recent conversations
 */
router.get("/api/chat/:userId/conversations", async (req, res) => {
  try {
    const { userId } = req.params;
    const { limit = 10 } = req.query;

    if (req.user.uid !== userId) {
      return res
        .status(403)
        .json({ error: "Cannot access other user's conversations" });
    }

    const conversations = await chatHistoryService.getRecentConversations(
      userId,
      parseInt(limit),
    );

    res.status(200).json(conversations);
  } catch (error) {
    logger.error(`Error retrieving conversations: ${error.message}`);
    res.status(500).json({ error: "Error retrieving conversations" });
  }
});

/**
 * GET /api/chat/:userId/conversation/:conversationId
 * Get a specific conversation
 */
router.get(
  "/api/chat/:userId/conversation/:conversationId",
  async (req, res) => {
    try {
      const { userId, conversationId } = req.params;

      if (req.user.uid !== userId) {
        return res
          .status(403)
          .json({ error: "Cannot access other user's conversation" });
      }

      const conversation = await chatHistoryService.getConversation(
        userId,
        conversationId,
      );

      if (!conversation) {
        return res.status(404).json({ error: "Conversation not found" });
      }

      res.status(200).json(conversation);
    } catch (error) {
      logger.error(`Error retrieving conversation: ${error.message}`);
      res.status(500).json({ error: "Error retrieving conversation" });
    }
  },
);

/**
 * DELETE /api/chat/:userId/history
 * Delete all chat history for user
 */
router.delete("/api/chat/:userId/history", async (req, res) => {
  try {
    const { userId } = req.params;

    if (req.user.uid !== userId) {
      return res
        .status(403)
        .json({ error: "Cannot delete other user's chat history" });
    }

    await chatHistoryService.deleteUserChatHistory(userId);

    res.status(200).json({ message: "Chat history deleted successfully" });
  } catch (error) {
    logger.error(`Error deleting chat history: ${error.message}`);
    res.status(500).json({ error: "Error deleting chat history" });
  }
});

/**
 * DELETE /api/chat/:userId/message/:messageId
 * Delete a specific message
 */
router.delete("/api/chat/:userId/message/:messageId", async (req, res) => {
  try {
    const { userId, messageId } = req.params;

    if (req.user.uid !== userId) {
      return res
        .status(403)
        .json({ error: "Cannot delete other user's message" });
    }

    await chatHistoryService.deleteMessage(userId, messageId);

    res.status(200).json({ message: "Message deleted successfully" });
  } catch (error) {
    logger.error(`Error deleting message: ${error.message}`);
    res.status(500).json({ error: "Error deleting message" });
  }
});

/**
 * POST /api/chat/:userId/search
 * Search user's chat history
 */
router.post("/api/chat/:userId/search", async (req, res) => {
  try {
    const { userId } = req.params;
    const { query } = req.body;

    if (req.user.uid !== userId) {
      return res
        .status(403)
        .json({ error: "Cannot search other user's chat history" });
    }

    if (!query) {
      return res.status(400).json({ error: "Search query is required" });
    }

    const results = await chatHistoryService.searchConversations(userId, query);

    res.status(200).json({ results, count: results.length });
  } catch (error) {
    logger.error(`Error searching chat history: ${error.message}`);
    res.status(500).json({ error: "Error searching chat history" });
  }
});

/**
 * GET /api/chat/:userId/statistics
 * Get user's chat statistics
 */
router.get("/api/chat/:userId/statistics", async (req, res) => {
  try {
    const { userId } = req.params;

    if (req.user.uid !== userId) {
      return res
        .status(403)
        .json({ error: "Cannot access other user's statistics" });
    }

    const stats = await chatHistoryService.getChatStatistics(userId);

    res.status(200).json(stats);
  } catch (error) {
    logger.error(`Error retrieving chat statistics: ${error.message}`);
    res.status(500).json({ error: "Error retrieving chat statistics" });
  }
});

/**
 * POST /api/chat/:userId/export
 * Export user's chat history
 */
router.post("/api/chat/:userId/export", async (req, res) => {
  try {
    const { userId } = req.params;
    const { format = "json" } = req.body;

    if (req.user.uid !== userId) {
      return res
        .status(403)
        .json({ error: "Cannot export other user's chat history" });
    }

    const exportedData = await chatHistoryService.exportChatHistory(
      userId,
      format,
    );

    // Set response headers for file download
    if (format === "json") {
      res.setHeader("Content-Type", "application/json");
      res.setHeader(
        "Content-Disposition",
        `attachment; filename="chat_history_${userId}_${Date.now()}.json"`,
      );
      res.send(JSON.stringify(exportedData, null, 2));
    } else if (format === "csv") {
      res.setHeader("Content-Type", "text/csv");
      res.setHeader(
        "Content-Disposition",
        `attachment; filename="chat_history_${userId}_${Date.now()}.csv"`,
      );
      res.send(exportedData);
    } else {
      return res
        .status(400)
        .json({ error: 'Invalid export format. Use "json" or "csv".' });
    }
  } catch (error) {
    logger.error(`Error exporting chat history: ${error.message}`);
    res.status(500).json({ error: "Error exporting chat history" });
  }
});

/**
 * GET /api/chat/:userId/reminders
 * Get academic reminders for user
 */
router.get("/api/chat/:userId/reminders", async (req, res) => {
  try {
    const { userId } = req.params;

    if (req.user.uid !== userId) {
      return res
        .status(403)
        .json({ error: "Cannot access other user's reminders" });
    }

    const reminders = await personalizationService.getAcademicReminders(userId);

    res.status(200).json(reminders);
  } catch (error) {
    logger.error(`Error retrieving reminders: ${error.message}`);
    res.status(500).json({ error: "Error retrieving reminders" });
  }
});

/**
 * GET /api/chat/:userId/personalization
 * Get user's personalization context
 */
router.get("/api/chat/:userId/personalization", async (req, res) => {
  try {
    const { userId } = req.params;

    if (req.user.uid !== userId) {
      return res
        .status(403)
        .json({ error: "Cannot access other user's personalization data" });
    }

    const context =
      await personalizationService.buildPersonalizationContext(userId);

    res.status(200).json(context);
  } catch (error) {
    logger.error(`Error building personalization context: ${error.message}`);
    res.status(500).json({ error: "Error building personalization context" });
  }
});

module.exports = router;

File: backend/functions/chat_routes.py
"""
backend/functions/chat_routes.py
Main chat endpoint with response mode integration
"""

from fastapi import APIRouter, HTTPException, Depends, status
from pydantic import BaseModel
from datetime import datetime
from typing import Optional, List, Dict, Any
import logging

from firebase import get_user_from_token, auth_required
from app.rag import answer_question
from app.services.firestore_service import FirestoreService
from app.services.chat_history_service import ChatHistoryService
from app.services.personalization_service import PersonalizationService

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/api", tags=["chat"])

# Initialize services
firestore_service = FirestoreService()
chat_history_service = ChatHistoryService()
personalization_service = PersonalizationService()


class UserPreferences(BaseModel):
    """User preferences for response formatting"""
    theme: Optional[str] = "system"
    responseLength: Optional[str] = "balanced"


class ChatRequest(BaseModel):
    """Request model for chat endpoint"""
    question: str
    responseMode: Optional[str] = "hybrid"  # "document", "web", or "hybrid"
    userId: str
    userPreferences: Optional[UserPreferences] = None
    conversationId: Optional[str] = None
    context: Optional[str] = None


class ChatResponse(BaseModel):
    """Response model for chat endpoint"""
    answer: str
    answer_source: str  # "document", "web_search", or "hybrid"
    sources: Optional[List[Dict[str, Any]]] = None
    web_sources: Optional[List[Dict[str, Any]]] = None
    responseTime: float
    reminders: Optional[List[Dict[str, str]]] = None
    confidence: Optional[float] = None


@router.post("/chat", response_model=ChatResponse)
@auth_required
async def chat_endpoint(
    chat_request: ChatRequest,
    current_user: dict = Depends(get_user_from_token)
):
    """
    Main chat endpoint with response mode support
    
    Args:
        chat_request: Chat request with question and mode
        current_user: Authenticated user from token
        
    Returns:
        ChatResponse: Answer with metadata
        
    Raises:
        HTTPException: If chat fails
    """
    import time
    start_time = time.time()
    
    # Verify user is sending their own request
    if current_user['uid'] != chat_request.userId:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Cannot process chat for other users"
        )

    try:
        # Get user profile for personalization
        user_profile = personalization_service.get_user_profile(chat_request.userId)
        
        # Build personalized context
        personalization_context = personalization_service.build_personalization_context(
            user_profile,
            chat_request.context
        )
        
        # Call RAG with response mode
        logger.info(f"Processing chat request - Mode: {chat_request.responseMode}, User: {chat_request.userId}")
        
        result = answer_question(
            question=chat_request.question,
            user_id=chat_request.userId,
            response_mode=chat_request.responseMode,  # Pass the selected mode
            personalization_context=personalization_context,
            response_length=chat_request.userPreferences.responseLength if chat_request.userPreferences else "balanced"
        )
        
        # Extract response details
        answer = result.get('answer', '')
        answer_source = result.get('answer_source', 'document')
        sources = result.get('sources', [])
        web_sources = result.get('web_sources', [])
        confidence = result.get('confidence', 0.0)
        
        # Get academic reminders if applicable
        reminders = []
        if chat_request.userPreferences and not chat_request.userPreferences.theme == "dark":
            reminders = personalization_service.get_academic_reminders(
                user_id=chat_request.userId,
                topic=chat_request.question
            )
        
        # Save conversation to history
        conversation_message = {
            "userId": chat_request.userId,
            "conversationId": chat_request.conversationId,
            "question": chat_request.question,
            "answer": answer,
            "answer_source": answer_source,
            "response_mode": chat_request.responseMode,
            "sources": sources,
            "web_sources": web_sources,
            "confidence": confidence,
            "timestamp": datetime.now().isoformat(),
            "userProfile": user_profile.get('id', 'unknown'),
            "responseTime": time.time() - start_time
        }
        
        await chat_history_service.save_message(conversation_message)
        
        # Log successful completion
        logger.info(f"Chat completed successfully - Source: {answer_source}, Time: {time.time() - start_time:.2f}s")
        
        return ChatResponse(
            answer=answer,
            answer_source=answer_source,
            sources=sources,
            web_sources=web_sources,
            responseTime=time.time() - start_time,
            reminders=reminders,
            confidence=confidence
        )
    
    except Exception as e:
        logger.error(f"Chat error: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error processing chat request: {str(e)}"
        )


@router.post("/chat/batch")
@auth_required
async def chat_batch_endpoint(
    requests: List[ChatRequest],
    current_user: dict = Depends(get_user_from_token)
):
    """
    Batch chat endpoint for multiple questions
    
    Args:
        requests: List of chat requests
        current_user: Authenticated user from token
        
    Returns:
        List[ChatResponse]: List of responses
    """
    results = []
    
    for req in requests:
        # Verify user ownership
        if current_user['uid'] != req.userId:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Cannot process chat for other users"
            )
        
        # Process each request
        response = await chat_endpoint(req, current_user)
        results.append(response)
    
    return results


@router.get("/chat/conversation/{conversation_id}")
@auth_required
async def get_conversation(
    conversation_id: str,
    current_user: dict = Depends(get_user_from_token)
):
    """
    Retrieve a specific conversation
    
    Args:
        conversation_id: The conversation ID
        current_user: Authenticated user from token
        
    Returns:
        dict: Conversation with all messages
    """
    try:
        conversation = chat_history_service.get_conversation(conversation_id)
        
        # Verify ownership
        if conversation.get('userId') != current_user['uid']:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Cannot access other user's conversations"
            )
        
        return conversation
    
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error retrieving conversation: {str(e)}"
        )


@router.get("/chat/recent")
@auth_required
async def get_recent_chats(
    limit: int = 10,
    current_user: dict = Depends(get_user_from_token)
):
    """
    Get recent chat conversations for user
    
    Args:
        limit: Maximum number of conversations to return
        current_user: Authenticated user from token
        
    Returns:
        List[dict]: Recent conversations
    """
    try:
        conversations = chat_history_service.get_recent_conversations(
            current_user['uid'],
            limit=limit
        )
        return conversations
    
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error retrieving recent chats: {str(e)}"
        )


@router.get("/chat/search")
@auth_required
async def search_chats(
    query: str,
    current_user: dict = Depends(get_user_from_token)
):
    """
    Search through user's chat history
    
    Args:
        query: Search query
        current_user: Authenticated user from token
        
    Returns:
        List[dict]: Matching conversations
    """
    try:
        results = chat_history_service.search_conversations(
            current_user['uid'],
            query
        )
        return results
    
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error searching chats: {str(e)}"
        )
